iscsiuio/src/unix/main.c

```diff
@@ -161,6 +193,28 @@ static void *signal_handle_thread(void *arg)
 
 	ILOG_INFO("terminating...");
 
+	/*
+	 * for debugging shutdown issues, let's wait 10 seconds, max,
+	 * to ensure all of our threads shutdown, since we have seen
+	 * issues where they may not do so.
+	 */
+	waitcount = WAITCOUNT_MAX;
+	while (waitcount--) {
+		sleep(1);
```

Comment:
Would moving the sleep to the end of the loop prevent taking a 1 second sleep when none is needed?
Do you have an idea of what the average wait time here is in your testing?  Just curious.

Reply:
I will look at moving the sleep(1). In answer to your question, we ran about 50 reboot cycles and never saw a delay of 2 seconds (with iSCSI root).

Reply:
@cleech what do you think about this?

```
        ...
	waitcount = WAITCOUNT_MAX;
	while ((event_loop_observers > 0) && waitcount--) {
		sleep(1);
		if (event_loop_observers <= 0)
			break;	/* they finished while we were sleeping */
		ILOG_INFO("%d threads still polling event_loop_stop flag after %d seconds",
			  event_loop_observers, WAITCOUNT_MAX - waitcount);
	}
	if (event_loop_observers < 0)
		ILOG_DEBUG("Invalid observer count: %d", event_loop_observers);
	else if (event_loop_observers > 0)
		ILOG_ERR("%d unresponsive observers will be cancelled: %d",
			 event_loop_observers);
	cleanup();
	exit(EXIT_SUCCESS);
```
This has the advantage that if there are no observers at all when we first enter the loop, we will not sleep at all. But, in practice, I think it's likely the sleep(1) that gives the threads a chance to complete their runs and decrease the observer count. So the code checks again after waking up. I also changed the message about a negative observer count to a debug message, so we don't freak normal users out.

I also think we could likely change the maximum sleep to something like 5 seconds, but in practice (as I said before) we only once saw a sleep of 2 seconds.

If you approve of the above suggested change, I'll push an update.